shared/route-tree/index.js

```diff
@@ -0,0 +1,229 @@
+  | { containerComponent: Component<any, P, any> }
+)
+
+export class RouteDefNode extends _RouteDefNode {
+  constructor ({defaultSelected, component, containerComponent, tags, initialState, children}: RouteDefParams<*>) {
+    super({
+      defaultSelected: defaultSelected || null,
+      component,
+      containerComponent,
+      tags: LeafTags(tags),
+      initialState: I.Map(initialState),
+      props: I.Map(),
+      state: I.Map(),
+      children: I.Seq(children)
+        .map(params => params instanceof RouteDefNode || typeof params === 'function' ? params : new RouteDefNode(params))
+        .toMap(),
+    })
+  }
+
+  getChild (name: string): ?RouteDefNode {
+    const childDef = this.children.get(name)
+    if (!childDef) {
+      return
+    }
+    if (typeof childDef === 'function') {
+      return childDef()
+    }
+    return childDef
+  }
+}
+
+type RouteStateParams = {
+  selected: string | null,
+  props?: I.Map<string, any>,
+  state?: I.Map<string, any>,
+}
+
+const _RouteStateNode = I.Record({
+  selected: null,
+  props: I.Map(),
+  state: I.Map(),
+  children: I.Map(),
+})
+
+export class RouteStateNode extends _RouteStateNode {
+  constructor (data: RouteStateParams) {  // eslint-disable-line no-useless-constructor
+    super(data)
+  }
+
+  getChild (name: string): RouteStateNode {
+    return this.children.get(name)
+  }
+
+  updateChild (name: string, op: (node: RouteStateNode) => ?RouteStateNode): RouteStateNode {
+    return this.updateIn(['children', name], op)
+  }
+}
+
+export class InvalidRouteError extends Error {}
+
+// Explicit list of iterable types to accept. We don't want to allow strings
+// since navigateTo('foo') instead of navigateTo(['foo']) is an easy mistake to
+// make.
+type PathIterable<X> = I.IndexedSeq<X> | I.List<X> | Array<X>
+export type Path = PathIterable<string>
+export type PropsPath = PathIterable<string | {selected: string | null}>
+type PathParam = [] | Path | PropsPath  // Flow doesn't accept Path as a subtype of PropsPath, so be explicit here.
+type PathSetSpec = I.IndexedIterable<{type: 'next' | 'navigate', next: string | null, props?: {}}>
+
+// Traverse a routeState making changes according to the pathSpec. This is the
+// primary mutation function for navigation and changing props. It will follow
+// the "next" props of each item in the pathSpec, then following any
+// defaultSelected from the routeDefs. It creates any routeState nodes that
+// don't exist along the way.
+function _routeSet (routeDef: RouteDefNode, routeState: ?RouteStateNode, pathSpec: PathSetSpec): RouteStateNode {
+  const pathHead = pathSpec && pathSpec.first()
+
+  let newRouteState = routeState || new RouteStateNode({selected: routeDef.defaultSelected})
+  if (pathHead && pathHead.type === 'navigate') {
+    newRouteState = newRouteState.set('selected', pathHead.next || routeDef.defaultSelected)
+    if (pathHead.next === null) {
+      newRouteState = newRouteState.delete('children')
```

Comment:
Ah! This could use a comment. We reset the state of children of a route when navigating to it. For example, if you navigate to `/settings`, we want the subnav to reset.